main/src/library/LazyList.flix

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2020 Esben Bjerre
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace LazyList {
+
+    ///
+    /// An immutable lazy list.
+    ///
+    /// A lazy list is either the empty list represented by `Empty`, or
+    /// an element `x` followed by a lazy list `xs`.
+    /// Note that both the head and tail of a lazy list is lazy.
+    ///
+    pub enum LazyList[t] {
+        case Empty,
+        case Cons(Lazy[t], Lazy[LazyList[t]])
+    }
```

Comment:
There are actually two types of lazy lists that I would like to consider, but neither is this one :)

The first is:

```
    pub enum LazyList[t] {
        case Empty,
        case Cons(t, Lazy[LazyList[t]])
    }
```

But why is the element `t` not lazy you ask? It turns out to be unnecessary, since I can do `List[Lazy[Int]]` or `LazyList[Lazy[Int]]`. The second definition -- lets call it LazyList2 for now -- captures the intuition that you want to have a list where not even the first element is evaluated:

```
    pub enum LazyList2[t] {  
      case LazyList2(Lazy[LList[t]])
    }
    pub enum InnerLazyList[t] {
        case Empty,
        case Cons(t, Lazy[InnerLazyList[t]])
    }
```

Reply:
Does that make sense? :)